### PR TITLE
Lower memory consumption by streaming request data

### PR DESCRIPTION
**Pull request by rtyley**
## Motivation and Context

Prior to this change, the `UrlConnectionHttpClient` implementation of `SdkHttpClient` would unintentionally cause **the entire `RequestBody`** to be loaded into memory before being sent, even if the body had been made using `software.amazon.awssdk.core.sync.RequestBody.fromInputStream()` and thus was a stream of known content-length. Even worse, as the buffering `ByteArrayInputStream` received the request, it would have to repeatedly grow capacity by allocating new arrays and copying the data over, so peak memory usage would be 1.5x the size of the request data. For instance, streaming 1 GB of data to S3 with `putObject` would lead to a peak memory allocation of 1.5 GB. This large allocation of memory can be avoided by instead *streaming* the request data as it is sent.

`java.net.HttpURLConnection` (the core Java class used by `UrlConnectionHttpClient` to implement `SdkHttpClient`) supports streaming the request, but will only do it if `setFixedLengthStreamingMode()` or `setChunkedStreamingMode()` has been called - this is true even if the `Content-Length` header has already been set. You can [see this in the implementation](https://github.com/openjdk/jdk/blob/da75f3c4ad5bdf25167a3ed80e51f567ab3dbd01/src/java.base/share/classes/sun/net/www/protocol/http/HttpURLConnection.java#L1373-L1394) of `sun.net.www.protocol.http.HttpURLConnection` where `streaming()` must be true to avoid **the whole request** being read into a `PosterOutputStream` (a simple extension of `ByteArrayInputStream`) before being sent.

## Description

If `Content-Length` is known, we're actually free to call `setFixedLengthStreamingMode()` on the connection to activate streaming, the only question is where & when to do it! Unfortunately `software.amazon.awssdk.http.urlconnection.UrlConnectionFactory`, which provides a way for end-users of the AWS SDK to customise the creation & configuration of the `HttpURLConnection`, _doesn't_ have access to the request or it's `Content-Length`, so can't call `setFixedLengthStreamingMode()`. In any case, this behaviour probably _always_ makes sense, so it doesn't make sense to leave it to users to customise it. That leaves the best place to do this being the `createAndConfigureConnection()` method, where the newly created `HttpURLConnection` is available and the `Content-Length` can be read if it's present.

Consequently, I've updated `createAndConfigureConnection()` to call `setFixedLengthStreamingMode` if the `Content-Length` is available and above a streaming threshold of 16 MB. It's possible that buffering an entire request in memory is beneficial in some way for short requests, I don't know, but I believe for large request objects avoiding the large memory allocation is definitely better.

## Testing

At present, the only way to workaround this issue without a change to the AWS SDK is to create a custom `SdkHttpClient` that wraps `UrlConnectionHttpClient` and overrides `prepareRequest()`. As the `connection` field is not visible in the `ExecutableHttpRequest` object (it's defined on the private `software.amazon.awssdk.http.urlconnection.UrlConnectionHttpClient.RequestCallable` class), it's necessary to use reflection to extract the `connection` field before operating on it. An example of this in Scala code can be seen here:

https://github.com/guardian/ophan-geoip-db-refresher/commit/e2fc370719b020178d30d5266b5363d8c2822311#diff-5780ade773d1bdf8dd5dea40283e6fec88bacdb30c20c5b39955da1faa8e3ade

Running the `ophan-geoip-db-refresher` code locally, the heap required is:

* 560 MB (`-Xmx560m`) - _without_ the workaround, request fully in memory
*  32 MB - with the workaround on, and so streaming enabled

## Screenshots

These videos show memory usage before and after streaming being enabled:

https://user-images.githubusercontent.com/52038/142726103-a347b8a1-e72d-425a-b98e-b1be0a301e59.mov

...at 31s in the above video, you can see memory usage suddenly massively increase - because the entire outgoing request has been buffered into memory. In the video below, with streaming enabled, memory consumption stays low for the entire duration of the run:

https://user-images.githubusercontent.com/52038/142726309-cd6b6579-c9c4-45fb-84af-6571c8e12545.mov

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [ ] Local run of `mvn install` - fails on main, fails in same way on this branch
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by zoewangg** [comment]
Hi @rtyley, apologies for the delayed response and thank you for your PR! We will take a look shortly

**Comment by rtyley** [comment]
> Hi @rtyley, apologies for the delayed response and thank you for your PR! We will take a look shortly

Thanks @zoewangg I appreciate it! I can rebase on the `master` branch to repair the `CHANGELOG.md` merge conflict if that helps?

File: http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java [hunk 1]
```diff
@@ -140,6 +142,10 @@ private HttpURLConnection createAndConfigureConnection(HttpExecuteRequest reques
         // See: https://github.com/aws/aws-sdk-java-v2/issues/975
         connection.setInstanceFollowRedirects(false);
 
+        request.httpRequest().firstMatchingHeader(CONTENT_LENGTH).map(Long::parseLong)
+               .filter(len -> len > STREAMING_OUTPUT_THRESHOLD)
+               .ifPresent(connection::setFixedLengthStreamingMode);
```
[the remark this answers is not in the document]
Comment (rtyley) [reply http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java:147]: Nice, thanks for the sample from `ApacheHttpRequestFactory`, I assumed that something similar was happening there.

> Is there any reason we are only using streaming if the length is larger than 16MB? I wonder if we should follow the same approach used in ApacheHttpClient, which is only buffering the content if the content length is not null.

I didn't know what `ApacheHttpClient` was doing but that makes perfect sense, yup, the 16MB threshold was completely arbitrary on my part - I'll take it out.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,7 @@
+## __URL Connection Http Client__
```
[the remark this answers is not in the document]
Comment (rtyley) [reply CHANGELOG.md]: Just ran it, thanks for the tip - I would note, that script made me install `vim`!

```
./scripts/new-change
/bin/sh: 1: vim: not found
Traceback (most recent call last):
  File "./scripts/new-change", line 229, in <module>
    main()
  File "./scripts/new-change", line 225, in main
    sys.exit(new_changelog_entry(args))
  File "./scripts/new-change", line 88, in new_changelog_entry
    parsed_values = get_values_from_editor(args)
  File "./scripts/new-change", line 123, in get_values_from_editor
    parsed_values = parse_filled_in_contents(filled_in_contents)
  File "./scripts/new-change", line 194, in parse_filled_in_contents
    raise Exception("Unsupported category %s" % t)
Exception: Unsupported category 
```

...but once that was done, it worked.

File: http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (rtyley) [reply http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java]: Threshold for streaming is removed with e1feaa6, thanks

File: http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java [hunk 3]
```diff
@@ -140,6 +141,9 @@ private HttpURLConnection createAndConfigureConnection(HttpExecuteRequest reques
         // See: https://github.com/aws/aws-sdk-java-v2/issues/975
         connection.setInstanceFollowRedirects(false);
 
+        request.httpRequest().firstMatchingHeader(CONTENT_LENGTH).map(Long::parseLong)
```
Comment (zoewangg) [inline http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java:144]: Do we need `.map(Long::parseLong)`?

Comment (rtyley) [reply http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java:144]: Yes, pretty sure we do - [`firstMatchingHeader(CONTENT_LENGTH)`](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/http/SdkHttpHeaders.html#firstMatchingHeader-java.lang.String-) returns an `Optional<String>`, and [`setFixedLengthStreamingMode()`](https://docs.oracle.com/javase/8/docs/api/java/net/HttpURLConnection.html#setFixedLengthStreamingMode-long-) accepts a `long`, so we need to map that `String` into a `long`.

Comment (zoewangg) [reply http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java:144]: Ah yeah, got it

**Review by zoewangg — APPROVED** [review]
Thank you for your contribution!

**Comment by zoewangg** [comment]
@all-contributors please add @rtyley for code.

**Pull request merged**